### DISCUSSION TITLE
Meeting 2024-03-21

### DISCUSSION THREAD

**Discussion by wellington36** [wellington36/stan_summer#3] [General]
# Structure
### stan/infiniteSumToThreshold.stan (31 lines)
 The function receives the parameters, epsilon, maxIter and n0 (start), when a_n is given by logFunction(n, p) and return de `log_sum_exp` of the a_n and the number of terms
 - 05-08 start variables
 - 10-13 first two terms
 - 15-21 evaluate the log(a_n) to the maximum
 - 23-28 evaluates log(a_n) until the condition log(a_n) > log(eps) is met.
 - 29-29 return the result with log_sum_exp and sort_asc.

### stan/infiniteErrorBoundingPairs.stan (51 lines)
 The function receives the parameters, epsilon, maxIter, log of L and n0, when $a_n$ is given by logFunction(n, p) and return the middle of the interval
 - 01-14 conditional stop criterion function
 - 20-25 start variables
 - 28-30 first two terms
 - 33-38 evaluate the log(a_n) to the maximum
 - 40-45 evaluates log(a_n) until the condition defined at the beginning is
 reached
 - 46-49 take the middle of the interval

### stan/InfiniteBatches.stan (41 lines)
 The function receive the parameters, batch_size, epsilon, maxIter and n0 (start), when a_n is given by logFunction(n, p) and return de log_sum_exp of the a_n and the number of terms
 - 05-13 start variables
 - 14-19 store the first batch and her log_sum_exp
 - 22-28 testing convergence and find the maximum
 - 30-37 evaluating the batches and adding their sum at the begin 

### stan/COMP_normalizing.stan (59 lines)
 The composition normalizing for test
 - 01-25 define functions (logFunction, true_value and difference_check)
 - 26-29 import the functions from approaches 1 and 2
 - 31-42 set data type
 - 44-57 apply the functions from approach 1 and 2 to the COMP

### R/test_comp_normalizing.R (38 lines)
 Test the COMP with the functions from approaches 1 and 2
 - 04-09 auxiliary functions
 - 11-13 load the model
 - 15-19 insert the parameters
 - 21-28 run MCMC
 - 30-36 print the results

### n_mixture/infiniteSumToThreshold.stan (31 lines)
 The function is equivalent to `stan/infiniteSumToThreshold.stan` with the difference the logFuncion receives one more argument `int[] ns`.

### n_mixture/n_mixtures.stan (45 lines)
 The n mixtures model to test
 - 01-10 define the logFunction and import the function from sumToThreshold
 - 12-22 set data type
 - 24-38 set parameters
 - 40-45 create the model

### R/test_n_mixtures.R (6 lines)
 Only load the `n_mixtures.stan` to check if works

# Questions/Sugestions
(1) In approach 1 we use M=1/2, this works well of L=0, but in other cases not. A sugestion is use M=(1+L)/2.
(2) In approach 2 on the paper, why do we check the condition `while a_{n+1} > a_n`? But in the code isn't checked (and looks unnecessary).
(3) About approach 2, in the file `stan/infiniteErrorBoundingPairs.stan` we have the function `adaptiveConvergenceCheck` that checks for convergence, it is not at all obvious that it is what is described in the paper (I tried to do check but got lost with the logarithms), perhaps to make it clearer.
(4) A suggestion to go now is to organize an example/test folder like the COMP normalizing maybe.
(5) The COMP only uses approaches 1 and 2, while the n_mixture uses approach 1. Perhaps we can utilize all approaches, but it is necessary to check if it is possible to do so.

@maxbiostat

**Comment by wellington36** [comment]
See the last changes https://github.com/maxbiostat/stan_summer/pull/1#issue-2194797997.

**Comment by maxbiostat** [comment]
> # Questions/Sugestions
> (1) In approach 1 we use M=1/2, this works well of L=0, but in other cases not. A sugestion is use M=(1+L)/2.

Nice suggestion, thanks. Without re-reading the manuscript, it's not immediately clear to me why `M=(L+1)/2` is a good choice. It'd be nice to have this explanation immortalised here. 

>  (2) In approach 2 on the paper, why do we check the condition `while a_{n+1} > a_n`? But in the code isn't checked (and looks unnecessary).

I believe this has to do with checking whether the series is now decreasing. We want to sum things that are _eventually_ decreasing, but they might grow up to a mode and then decrease. I think the implementation handles 'finding the maximum' differently and that's why this particular condition is not checked. 

> (3) About approach 2, in the file `stan/infiniteErrorBoundingPairs.stan` we have the function `adaptiveConvergenceCheck` that checks for convergence, it is not at all obvious that it is what is described in the paper (I tried to do check but got lost with the logarithms), perhaps to make it clearer.

Sure. We'll sort that out when the manuscript is ready for resubmission. @GuidoAMoreira has a peculiar coding style 😆. 

>  (4) A suggestion to go now is to organize an example/test folder like the COMP normalizing maybe. (5) The COMP only uses approaches 1 and 2, while the n_mixture uses approach 1. Perhaps we can utilize all approaches, but it is necessary to check if it is possible to do so.

Yes. Ideally all examples would use all approaches. Let's update this.

**Reply by wellington36** [reply]
>Nice suggestion, thanks. Without re-reading the manuscript, it's not immediately clear to me why M=(L+1)/2 is a good choice. It'd be nice to have this explanation immortalised here.

The choice is based on the fact that M belongs to the interval (L, 1), if M is close to L, then $n_0$ is large, which makes the method's validity difficult. If M is close to 1, $U_\varepsilon = \varepsilon \frac{M}{1 - M} \to \infty$, that is, the guaranteed approximation is worse.

For a fixed value, if M =1/2 (in our case) is close to L, then $n_0 \to \infty$, which means that the condition $n > n_0$ does not hold.

So a good choice would be not to take M close to the edges of the interval, thus being the midpoint. I had recommended this at the SEEMAp presentation last year (in which I discussed sumToThreshold and some examples):

![image](https://github.com/wellington36/stan_summer/assets/61877847/e04fb711-e2fb-4b97-98f9-2c0b1104f9dd)
_SEEMAp slideshow – October 2023_
